### ISSUE TITLE
Process on Windows is vulnerable to BatBadBut

### ISSUE THREAD

**Issue by maia-s** [libsdl-org/SDL#10857]
`CreateProcess*` uses different rules for escaping arguments when executing `bat` and `cmd` files. This earned several programming languages a critical CVE back in April because they didn't account for this in their standard libraries use of `CreateProcess`

See https://flatt.tech/research/posts/batbadbut-you-cant-securely-execute-commands-on-windows/ for details

Also, Windows strips trailing `.` and spaces from paths, which lead to a vulnerability in Rust's mitigation for this that was fixed recently

**Comment by slouken** [comment]
Thanks for the heads up! Do you have a proposed PR for this?

**Comment by icculus** [comment]
So to fix this, we would need to escape the arguments differently if it's a .bat/.exe file vs a .cmd file, but we don't _know_ it's a .cmd if the extension wasn't specified and the script is in the `%PATH%`, right?

It honestly seems like it would be easier to strip `.CMD` out of `%PATHEXT% before spawning the process. They'd still work if the extension is explicitly included, where it would be easier for us to know to escape it differently.

**Comment by maia-s** [comment]
To fix it, we'll have to fix the escaping for all batch files. The linked article says that includes "`.bat, .cmd, etc.`", but I don't know what that `etc` involves. If I understand correctly, `.exe` files can technically differ in how they do argument parsing too, but in practice they follow the convention currently implemented in SDL. (If not, there's no way for SDL to fix it anyway, so I think it's fair to call that a bug in the exe)

There's two cases: Either the program to be executed has an extension (after stripping trailing periods and spaces), or it does not.
- If it has an extension, we know if it's a batch file or not (assuming we know all affected extensions; at least `.bat` and `.cmd`, case insensitive), so we can apply the correct kind of escaping
- If it doesn't have an extension, we don't know. I don't know what's the right thing to do in this situation. Maybe filtering them out like icculus said would work, if that's applicable to all kinds of batch files (I don't know enough about Windows to answer this). Alternately, launching a process on Windows could fail if there's no extension, or SDL could append `.exe` if there's no extension, making the extension required only for other kinds of executables

(There's also the option to not fix this and just warn about it in the documentation, but that's prone to error)

> Thanks for the heads up! Do you have a proposed PR for this?

I don't have a PR. I could make one, but I'm not too familiar with low level Windows programming or details like the above, so I'd need input from someone who is.

**Comment by castholm** [comment]
Preface: This is just my own understanding from having experimented with `CreateProcessW` and command-line serialization in the past, so please don't blindly take my word for anything and double-check/verify things independently before taking action.

SDL might not be as badly affected by the BatBadBut vulnerability as some of the standard libraries mentioned in the blog post.

My understanding of the situation is as follows:

- `CreateProcessW` supports running programs with filename extensions ending in `.exe`, `.com`, `.bat` and `.cmd` (case-insensitive match).
- If the filename ends in `.bat` or `.cmd`, `CreateProcessW` rewrites the command line string such that `cmd.exe` is used to launch the batch file.
- If the specified program is not found/not supported, `CreateProcessW` internally appends `.exe` to the program path, then tries again.

In other words, `CreateProcessW` itself does not care for the `%PATHEXT%` env variable, and it does not append `.bat` or `.cmd` or any other extension to a path without a filename extension.

The reason Rust and some other ecosystems were so badly affected is because their Windows implementations of their process APIs tried to be a bit more helpful and replicate how the Windows command shell finds programs, which among other things involves appending the extensions listed in `%PATHEXT%` to the path.

But since SDL's Windows implementation of the process API is currently a thin wrapper over `CreateProcessW` and does not do any of these fancy things, it should be mostly safe, and is only affected by the issue when

- the user explicitly launches a `.bat` or `.cmd` file, or when
- the user explicitly launches `cmd.exe`.

SDL could be nice, try to manually detect `.bat` and `.cmd` files and escape them per the recommended BatBadBut mitigation strategies. But it might also be an okay resolution to simply add a doc comment to the API that clearly warns users about the dangers of launching `.bat`/`.cmd` files or `cmd.exe` on Windows, perhaps with a link to the BatBadBut article.

**Comment by madebr** [comment]
I added a reproducer to https://github.com/libsdl-org/SDL/pull/10836.

As a fix, how about the way [nodejs handles this](https://nodejs.org/en/blog/vulnerability/april-2024-security-releases-2#command-injection-via-args-parameter-of-child_processspawn-without-shell-option-enabled-on-windows-cve-2024-27980---high)?
If we detect the first argument expands to a .cmd or .bat, we refuse running the process.
Then, at a later time, we might add a`SDL_PROP_PROCESS_CREATE_SHELL_BOOLEAN` property.

**Comment by slouken** [comment]
> As a fix, how about the way [nodejs handles this](https://nodejs.org/en/blog/vulnerability/april-2024-security-releases-2#command-injection-via-args-parameter-of-child_processspawn-without-shell-option-enabled-on-windows-cve-2024-27980---high)? If we detect the first argument expands to a .cmd or .bat, we refuse running the process. Then, at a later time, we might add a`SDL_PROP_PROCESS_CREATE_SHELL_BOOLEAN` property.

No, if CreateProcess() launches it, we shouldn't prevent it if we can help it.
